src/controller/python/chip/storage/delegate.py

```diff
@@ -0,0 +1,131 @@
+import ctypes
+import chip.native
+import base64
+import threading
+from urllib import parse as urlparse
+
+@dataclass
+class StorageOptions:
+    path: str = ":memory:"
+    params: dict = field(default_factory=dict)
+
+# typedef void (*GetKeyValueFunct)(const uint8_t * key, void * value, uint16_t * size);
+# typedef void (*SetKeyValueFunct)(const uint8_t * key, const void * value, uint16_t size);
+# typedef void (*DeleteKeyValueFunct)(const uint8_t * key);
+_ControllerGetKeyValueFunct = CFUNCTYPE(None, c_char_p, c_size_t, c_void_p, ctypes.POINTER(c_uint16))
+_ControllerSetKeyValueFunct = CFUNCTYPE(None, c_char_p, c_size_t, c_void_p, c_uint16)
+_ControllerDeleteKeyValueFunct = CFUNCTYPE(None, c_char_p, c_size_t)
+
+_CONTROLLER_STORAGE_DOMAIN = "controller"
+_CONTROLLER_STORAGE_FIELD = "default"
+
+class SqlStorage:
+    def __init__(self, options: StorageOptions):
+        self.lock = threading.Lock()
+        self.con = sqlite3.connect(options.path, check_same_thread=False)
+        self.cache = dict()
+        cur = self.con.cursor()
+        cur.executescript('''
+        CREATE TABLE IF NOT EXISTS storage (
+            domain text NOT NULL,
+            field text NOT NULL,
+            key text NOT NULL,
+            value text NOT NULL
+        );
+        CREATE UNIQUE INDEX IF NOT EXISTS storage_index ON storage (
+            domain, field, key
+        );
+        ''')
+        self.con.commit()
+
+    def Set(self, domain:str, field:str, key:str, value:bytes)->None:
+        with self.lock:
+            cur = self.con.cursor()
+            valueStr = base64.standard_b64encode(value).decode('utf-8')
+            cur.execute('INSERT OR REPLACE INTO storage (domain, field, key, value) VALUES (?, ?, ?, ?)', (domain, field, key, valueStr))
+            self.con.commit()
+
+    def Get(self, domain:str, field:str, key:str)->bytes:
+        with self.lock:
+            cur = self.con.cursor()
+            cur.execute("SELECT value FROM storage WHERE domain = ? AND field = ? AND key = ? LIMIT 1", (domain, field, key))
+            res = cur.fetchone()
+            return base64.standard_b64decode(res[0]) if res is not None else None
+
+    def Delete(self, domain:str, field:str, key:str)->None:
+        with self.lock:
+            cur = self.con.cursor()
+            cur.execute("DELETE FROM storage WHERE domain = ? AND field = ? AND key = ?", (domain, field, key))
+            self.con.commit()
+
+_controllerStorage = None
+
+@_ControllerGetKeyValueFunct
+def ControllerGetKeyValueFunct(key:c_char_p, keyLen:int, valPtr:c_void_p, sizePtr:ctypes.POINTER(c_uint16)):
+    global _controllerStorage
+    accessKey = ctypes.string_at(key, keyLen).decode("utf-8")
+    data = _controllerStorage.Get(_CONTROLLER_STORAGE_DOMAIN, _CONTROLLER_STORAGE_FIELD, accessKey)
+    sizeRead = ctypes.pointer(c_uint16(len(data))) if data is not None else ctypes.pointer(c_uint16(0xffff))
```

Comment:
sizeRead is size of data - if not available, it seems that we want to start with 0 (as no data buffer is available) instead of 0xFFFF which basically means 'unlimited'.